### PR TITLE
Add initial query builders

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

In order to implement structured querying for the Graph API we chose to use a query builder approach to build our queries on the Rust side. We had a few requirements for our builders: They should be composable and deserializable, so it's possible to pass a full query from e.g. the REST API and directly translate it to Rust. This is the reason, why builders implemented in this PR do not contain a generic state argument.

This PR only adds very basic builders, which are able to specify the same functionality we had before.

## 🔗 Related links

- [Notion draft on structural querying](https://www.notion.so/hashintel/What-is-Structural-Querying-and-why-does-everyone-talk-about-it-9acbcd334c7f4d8d857b36e6ea81d89c) (_internal_)
- [Asana task](https://app.asana.com/0/0/1202770062090891/f) _(internal)_

## 🔍 What does this change?

- Add basic queries for
  - Ontology types (`DataType`, `PropertyType`, `LinkType`, `EntityType`)
  - `PersistedEntity`
  - `Links`
- Change the `Read` trait to use the queries
  - Currently, queries only need to be `Debug` (for logging) and `Sync`, there is no `Query` trait
  - `Read` is not bound to a lifetime anymore, `Query` will be passed by reference in any case (even though, the current queries could implement `Copy`, but this may change later and I chose to be conservative here)
  - `Read<T>::read` will *always* return `Vec<T>`
    - If we need different outputs, we can add this as a generic later
- Add the `Read` bound on the stores directly, e.g. `DataTypeStore: Read<DataType>` or `EntityStore: Read<PersistedEntity>`
- Remove the trait alias on `Graph` (which was only used to alias `Store` and the `Read` bounds)
- Only the previously implemented queries are implemented on queries, this means, that the graph will panic if an unsupported query is used (e.g. `DataType::with_version(10)`, this would read *all* `DataType`s with a version of `10`, but without specifying the URI).

## 📜 Does this require a change to the docs?

The external API does not change

## ⚠️ Known issues

- Queries always has the same output. This has two issues:
  - This is annoying when having queries like "give me entity with ID X and version Y", this will at most return one entity, but the `Read` trait returns a `Vec`. This could be resolved by adding a `read_one` method on the trait, which reuses the `read` method by default, but this is not required for now, so this was left out.
  - It will always return the most generic type, e.g. `PersistedEntity`. If only an account id ("Give me the account id for entity X") is required, the full entity is returned. This can be avoided later on by adding an additional generic to the `Read` trait, but to avoid overengineering now, this was left out. It's possible to add this functionality later.
- ~~Currently, the implementation of `Read` are pretty big because all the functionality is in one function, this should be moved out to dedicated functions in a follow-up.~~

## 🛡 What tests cover this?

The external API did not change, so the REST endpoint tests and the integration tests covers this. The integration test suite was slightly adjusted to create the queries in the API wrapper.

## ❓ How to test this?

Run the usual test suites.
I can highly recommend reviewing this offline, e.g. with CLion, as the diff-viewer is smarter there.

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/api/rest/api_resource.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use crate::GraphPool;
 /// through a `Router`, making it explicitly clear we want to provide `OpenApi` specification as
 /// documentation for the routes.
 pub(super) trait RoutedResource: utoipa::OpenApi {
-    fn routes<P: GraphPool>() -> Router;
+    fn routes<P: StorePool + Send + 'static>() -> Router;
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/api/rest/api_resource.rs:12]: `GraphPool` was an alias to `StorePool` with `StorePool::Store` bound to all the `Read` implementations, which is not needed anymore. `Send + 'static` is required for `axum`

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs [hunk 2]
```diff
@@ -139,13 +140,19 @@ async fn get_latest_data_types<P: GraphPool>(
...
-        .await
-        .map(Json)
+    read_from_store(
+        pool.as_ref(),
+        &DataTypeQuery::new()
+            .with_uri(uri.base_uri())
+            .with_version(uri.version()),
+    )
+    .await
+    .and_then(|mut data_types| data_types.pop().ok_or(StatusCode::NOT_FOUND))
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:154]: As mentioned in *Known issues*, `Read<T>::read` will now always return `Vec<T>`

File: packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs [hunk 3]
```diff
@@ -52,35 +49,28 @@ fn api_documentation() -> Vec<openapi::OpenApi> {
...
         tracing::error!(error=?report, "Could not acquire access to the store");
         StatusCode::INTERNAL_SERVER_ERROR
     })?;
 
-    crud::Read::<I, T>::get(&store, identifier)
-        .await
-        .map_err(|report| {
-            tracing::error!(error=?report, identifier=?identifier, "Could not read from the store");
-
-            if report.contains::<QueryError>() {
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/api/rest/mod.rs:74]: This test was moved to the caller function. For get-all statements, an empty list is valid, for get-one statements there will be an error on an empty list.

File: packages/graph/hash_graph/lib/graph/src/ontology/mod.rs [hunk 4]
```diff
@@ -9,8 +9,7 @@ use tokio_postgres::types::{FromSql, ToSql};
 use utoipa::Component;
 use uuid::Uuid;
 
-// TODO - find a good place for AccountId and VersionId, perhaps they will become redundant in a
-//  future design
+// TODO - find a good place for AccountId, perhaps it will become redundant in a future design
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/ontology/mod.rs:12]: Drive-by: `VersionId` is not here anymore

File: packages/graph/hash_graph/lib/graph/src/ontology/types/mod.rs [hunk 5]
```diff
@@ -26,8 +25,3 @@ pub use crate::ontology::types::{
 pub mod error;
 
 mod serde_shared;
-
-pub trait OntologyType {
-    /// Returns the unique versioned URI used to identify this instance of a type.
-    fn uri(&self) -> &VersionedUri;
-}
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/ontology/types/mod.rs:33]: Moved to `OntologyDatabaseType`, this trait was unused in a generic context outside of the database. For concrete types, `.id()` is available. `uri()` was confusing with the `uri()` function on `OntologyQuery`

File: packages/graph/hash_graph/lib/graph/src/store/mod.rs [hunk 6]
```diff
@@ -181,7 +187,7 @@ pub trait Store =
 
 /// Describes the API of a store implementation for [`DataType`]s.
 #[async_trait]
-pub trait DataTypeStore {
+pub trait DataTypeStore: for<'q> crud::Read<DataType, Query<'q> = DataTypeQuery<'q>> {
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/mod.rs:190]: As mentioned above, a store now has the crud-bounds directly.

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs [hunk 7]
```diff
@@ -6,81 +6,62 @@ use postgres_types::ToSql;
...
+                        INNER JOIN ids ON ids.version_id = entities.entity_type_version_id
+                        ORDER BY entity_id, entities.version DESC;
+                        "#,
+                        // Requires a concrete type, which implements
+                        // `IntoIterator<Item = impl BorrowToSql>`
+                        [] as [&(dyn ToSql + Sync); 0],
+                    )
+                    .await
+            }
+            _ => todo!(),
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:59]: The plan is to have a general query, that returns basically everything and filter at the Rust side. This is a solution to avoid having a permutation of querys for SQL. This probably can be optimized later, but for the near future this will work just fine.
This PR does not implement the logic because the goal of this PR is to only provide the queries without a breaking change to the REST API. It's not possible to hit this match arm from the REST API.

File: packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        self.version = Some(OntologyVersion::Exact(version));
+        self
+    }
+
+    #[must_use]
+    pub const fn with_latest_version(mut self) -> Self {
+        self.version = Some(OntologyVersion::Latest);
+        self
+    }
+}
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs:61]: Instead of `with_*` we probably want to use `by_*`. This enables us to add `with_*` methods later to change the return type.
We need to create the query (`by_x().by_y().by_z()`) without changing any generic parameters, as this is required for deserialization of a query. However, `with_*` may introduce additional parameters, which then can be used to determine the return type (e.g. `DataTypeQuery::new().by_uri(...).by_latest_version().with_account_id()` will return the account id of the specified URI+version)

**Review by TimDiekmann — COMMENTED** [review]
From conversations with @thehabbos007 and @Alfred-Mountfield

File: packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/read.rs [hunk 9]
```diff
@@ -5,86 +5,67 @@ use postgres_types::ToSql;
...
-            .await
-            .into_report()
-            .change_context(QueryError)
-            .attach_printable_lazy(|| uri.clone())?;
-
-        serde_json::from_value(row.get(0))
-            .into_report()
-            .change_context(QueryError)
-    }
+fn parameter_list<const N: usize>(list: [&(dyn ToSql + Sync); N]) -> [&(dyn ToSql + Sync); N] {
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/read.rs:15]: I suppose this is to get rid of the `[] as [&(dyn ToSql + Sync); 0]` we had previously in zero-arg queries?

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs [hunk 10]
```diff
@@ -6,81 +6,62 @@ use postgres_types::ToSql;
...
-            )
-            .await
-            .into_report()
-            .change_context(QueryError)?;
+    async fn read<'query>(
+        &self,
+        query: &Self::Query<'query>,
+    ) -> Result<Vec<PersistedEntity>, QueryError> {
+        let row_stream = match (query.id(), query.version()) {
+            (Some(entity_id), Some(EntityVersion::Latest)) => {
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:27]: I'm a little worried about the maintainability of this, it seems less than ideal that we're explicitly managing permutations.

I'm fine with us merging as is and trying to keep an eye on it though and see how it develops as we add more constraints to the queries.

File: packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    #[must_use]
+    pub const fn with_version(mut self, version: u32) -> Self {
+        self.version = Some(OntologyVersion::Exact(version));
+        self
+    }
+
+    #[must_use]
+    pub const fn with_latest_version(mut self) -> Self {
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs:57]: Should this `with_latest_version` (which i guess will become `by_latest_version`) not be different than the others? Instead of being a filter, this is more like an ordering + limiting clause. Perhaps `pick_latest_version` or something would be more appropriate.

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs (hunk 7 above)
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:59]: ^ See above about my concern 🙃 

Perhaps we can't really get around this without trying to generate SQL from some other intermediate representation, which also seems like a _**very**_ bad idea right now

File: packages/graph/hash_graph/tests/integration/postgres/mod.rs [hunk 12]
```diff
@@ -252,16 +298,34 @@ impl DatabaseApi<'_> {
...
+                    .with_source_entity_id(source_entity)
+                    .with_link_types(|link_types| {
+                        link_types
+                            .with_uri(link_type_uri.base_uri())
+                            .with_version(link_type_uri.version())
+                    }),
+            )
+            .await?
+            .pop()
+            .expect("No links found")
```
Comment (thehabbos007) [inline packages/graph/hash_graph/tests/integration/postgres/mod.rs:317]: Would be only a single link
```suggestion
            .expect("No link found")
```

**Review by thehabbos007 — CHANGES_REQUESTED** [review]
I have some questions and a couple change requests

**Review by Alfred-Mountfield — COMMENTED** [review]
Haven't done a full review but here are some comments

File: packages/graph/hash_graph/lib/graph/src/store/crud.rs [hunk 13]
```diff
@@ -3,35 +3,28 @@
...
-//   aware of the types.
-//   As an example why this is bad see the `AllLatest` marker struct.
+///
+/// [`Store`]: crate::store::Store
 #[async_trait]
-pub trait Read<'i, I: Send + 'i, T>: Store {
-    /// Output returned when getting the value by the identifier `I`.
-    type Output;
+pub trait Read<T: Send>: Sync {
+    // TODO: Implement `Valuable` for queries and use them directly
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/crud.rs:21]: It's a trait supported by `tracing` to print arguments. It's implemented for common types like `u16` and it's possible to implement it for custom types as well. This is an experimental tracing feature that I tested locally, but it's out of scope to do it properly now. The Debug output is fine for now I think.

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs [hunk 14]
```diff
@@ -77,7 +78,7 @@ struct CreateDataTypeRequest {
     ),
     request_body = CreateDataTypeRequest,
 )]
-async fn create_data_type<P: GraphPool>(
+async fn create_data_type<P: StorePool + Send>(
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:81]: Yeah, that's a bit unfortunate, but at least it's only a single bound. I still prefer it over a trait alias just for this purpose. I don't want to add the bound to `StorePool` directly as this is unrelated to `StorePool` and only required by `axum`.

File: packages/graph/hash_graph/lib/graph/src/api/rest/property_type.rs [hunk 15]
```diff
@@ -115,12 +117,15 @@ async fn create_property_type<P: GraphPool>(
...
     )
 )]
-async fn get_latest_property_types<P: GraphPool>(
+async fn get_latest_property_types<P: StorePool + Send>(
     pool: Extension<Arc<P>>,
 ) -> Result<Json<Vec<PropertyType>>, StatusCode> {
-    read_from_store::<PropertyType, _, _, _>(pool.as_ref(), AllLatest)
-        .await
-        .map(Json)
+    read_from_store(
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/property_type.rs:123]: > Turbo-swordfish

That's new 🤣

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 16]
```diff
@@ -3,118 +3,137 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
...
-            .change_context(QueryError)
-            .attach_printable(identifier)?
-            .into_iter()
-            .map(|row| (VersionedUri::new(row.get(0), row.get::<_, i64>(1) as u32), Outgoing::Multiple(row.get(2))));
+    #[expect(
+        clippy::too_many_lines,
+        reason = "TODO: Function needs to be split into smaller parts"
+    )]
+    async fn read<'query>(&self, query: &Self::Query<'query>) -> Result<Vec<Links>, QueryError> {
+        match (query.link_type_query(), query.source_entity_id()) {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:25]: Sure, can do it now. I will split this up.

File: packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/read.rs (hunk 9 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/read.rs:15]: Yes, and it's also useful to convert other types to convert to the desired type. It's convenient to call it when interacting with `query_raw`.

File: packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs (hunk 11 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs:57]: I wonder if we want to explore this a bit more (genuinely maybe not wort hit right now), I could see us wanting other kinds of queries where we're saying "pick the first/last element when sorted by X". I think this is still a sort of filter, but as you're saying I think it's more a combination of two operations, ones that we don't necessarily have good plans to support yet, or know how to handle them.

File: packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/api/rest/data_type.rs:154]: It should definitely be possible but doing this now is out of the scope of this PR:
> This could be resolved by adding a read_one method on the trait, which reuses the read method by default, but this is not required for now, so this was left out.

I will revisit this at a later point. I added a todo for this: ebdb86b44f4ab4bf298017c84cfa9a7ae90a1b25

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs (hunk 10 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:66]: I added a related comment here: https://github.com/hashintel/hash/pull/909#discussion_r942116961

I'm also not super happy with this currently, but for now, this is working fine. I will keep this in mind when further extending the queries.

File: packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs (hunk 8 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/query/ontology.rs]: I renamed `with_*` to `by_*` in 202bd3266fcd5923071ccbaef3e9bd0088307e83.

File: packages/graph/hash_graph/lib/graph/src/store/mod.rs [hunk 17]
```diff
@@ -242,14 +246,11 @@ pub trait PropertyTypeStore {
     /// # Errors
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/mod.rs:246]: Comment on line 244 shouldn't have `identifier`. I think this is the case for all the `get_*_type` methods as well as `get_entity`

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs [hunk 18]
```diff
@@ -3,82 +3,72 @@ use chrono::{DateTime, Utc};
...
+    client
+        .query_raw(
+            r#"
+            SELECT DISTINCT ON(entity_id) properties, entity_id, entities.version, ids.base_uri, ids.version, created_by
+            FROM entities
+            INNER JOIN ids ON ids.version_id = entities.entity_type_version_id
+            ORDER BY entity_id, entities.version DESC;
+            "#,
+            // Requires a concrete type, which implements `IntoIterator<Item = impl BorrowToSql>`
+            [] as [&(dyn ToSql + Sync); 0],
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:51]: Couldn't we use the `parameter_list` utility function across `knowledge` and `ontology`? So we can replace this as well

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs [hunk 19]
```diff
@@ -3,82 +3,72 @@ use chrono::{DateTime, Utc};
...
-
-        let entity_id: EntityId = row.get(1);
-        let entity_version: DateTime<Utc> = row.get(2);
-        let type_base_uri: String = row.get(3);
-        let type_version: i64 = row.get(4);
-        let created_by: AccountId = row.get(5);
+                WHERE entity_id = $1
+            );
+            "#,
+            &[&entity_id],
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:34]: Here `parameter_list` can be utilized as well

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 20]
```diff
@@ -3,118 +3,163 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
...
+) -> Result<impl Iterator<Item = (VersionedUri, Outgoing)> + Send, QueryError> {
+    Ok(client
+        .query(
+            r#"
+            SELECT base_uri, "version", target_entity_id
+            FROM links
+            INNER JOIN ids ON ids.version_id = links.link_type_version_id
+            WHERE active AND NOT multi and source_entity_id = $1
+            "#,
+            &[&source_entity_id],
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:27]: Here `parameter_list` can be utilized as well

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 21]
```diff
@@ -3,118 +3,163 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
...
-                    Outgoing::Single(row.get(2)),
-                )
-            });
-
-        Ok(Links::new(multi_links.chain(single_links).collect()))
-    }
+            SELECT base_uri, "version", links FROM aggregated
+            INNER JOIN ids ON ids.version_id = aggregated.link_type_version_id
+            "#,
+            &[&source_entity_id],
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:58]: Here `parameter_list` can be utilized as well

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 22]
```diff
@@ -3,118 +3,163 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
...
+                WHERE active AND multi AND source_entity_id = $1 AND base_uri = $2 AND "version" = $3 GROUP BY link_type_version_id
+            )
+            -- Combine single and multi links with null values in rows where the other doesn't exist
+            SELECT link_type_version_id, target_entity_id AS single_link, NULL AS multi_link
+            FROM single_links
+            UNION
+            SELECT link_type_version_id, NULL AS single_link, target_entity_ids AS multi_link
+            FROM multi_links
+            "#,
+            &[&source_entity_id, link_type_uri, &i64::from(link_type_version)],
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:111]: Here `parameter_list` can be utilized as well

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs [hunk 23]
```diff
@@ -3,118 +3,163 @@ use error_stack::{IntoReport, Report, Result, ResultExt};
...
+    Ok(vec![Links::new(
+        [(
+            VersionedUri::new(link_type_uri.to_string(), link_type_version),
+            outgoing,
+        )]
+        .into(),
+    )])
+}
+
+// TODO: we should probably support taking PersistedEntityIdentifier here as well as an EntityId
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:138]: Not entirely sure what this comment refers to? 
Doe sit mean that in place of having the `LinkQuery` specify a `source_entity_id` which is of type `EntityId`, we can tak in `PersistedEntityIdentifier`?

**Review by thehabbos007 — COMMENTED** [review]
Couple more things here

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs (hunk 20 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:27]: Not possible here because it is using `query` rather than `query_raw`

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs (hunk 21 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:58]: Same as above

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs (hunk 22 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:111]: Not possible here because it is using `query_one` rather than `query_raw`

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs (hunk 18 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs]: Applied in eefbbb505466911cf5d6ff78dd0ac2e8945eb546

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs (hunk 19 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs]: Applied in eefbbb505466911cf5d6ff78dd0ac2e8945eb546

File: packages/graph/hash_graph/lib/graph/src/store/mod.rs (hunk 17 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/mod.rs:246]: Applied in 4dfc8ec7ba2081be1564694186602ff6cc00b748

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs (hunk 23 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/link/read.rs:138]: Yes 🙂

**Review by thehabbos007 — APPROVED** [review]
:rocket: 
~ :rock: ~
